compactcert/builder.go

```diff
@@ -0,0 +1,332 @@
+	if err != nil {
+		return builder{}, err
+	}
+
+	ccw.builders[rnd] = res
+	return res, nil
+}
+
+func (ccw *Worker) initBuilders() {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		roundSigs, err = getPendingSigs(tx)
+		return
+	})
+	if err != nil {
+		ccw.log.Warnf("initBuilders: getPendingSigs: %v", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		_, ok := ccw.builders[rnd]
+		if ok {
+			ccw.log.Warnf("initBuilders: round %d already present", rnd)
+			continue
+		}
+
+		builder, err := ccw.builderForRound(rnd)
+		if err != nil {
+			ccw.log.Warnf("initBuilders: builderForRound(%d): %v", rnd, err)
+			continue
+		}
+
+		for _, sig := range sigs {
+			pos, ok := builder.voters.AddrToPos[sig.signer]
+			if !ok {
+				ccw.log.Warnf("initBuilders: cannot find %v in round %d", sig.signer, rnd)
+				continue
+			}
+
+			err = builder.Add(pos, sig.sig, false)
+			if err != nil {
+				ccw.log.Warnf("initBuilders: cannot add %v in round %d: %v", sig.signer, rnd, err)
+				continue
+			}
+		}
+	}
+}
+
+func (ccw *Worker) handleSigMessage(msg network.IncomingMessage) network.OutgoingMessage {
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		ccw.log.Warnf("ccw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := ccw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		ccw.log.Warnf("ccw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+func (ccw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	builder, ok := ccw.builders[sfa.Round]
+	if !ok {
+		latest := ccw.ledger.Latest()
+		latestHdr, err := ccw.ledger.BlockHdr(latest)
+		if err != nil {
+			return network.Disconnect, err
+		}
+
+		if sfa.Round <= latestHdr.CompactCertLastRound {
+			// Already have a complete compact cert in ledger.
+			// Ignore this sig.
+			return network.Ignore, nil
+		}
+
+		builder, err = ccw.builderForRound(sfa.Round)
+		if err != nil {
+			return network.Disconnect, err
+		}
+	}
+
+	pos, ok := builder.voters.AddrToPos[sfa.Signer]
+	if !ok {
+		return network.Disconnect, fmt.Errorf("handleSig: %v not in participants for %d", sfa.Signer, sfa.Round)
+	}
+
+	if builder.Present(pos) {
+		// Signature already part of the builder, ignore.
+		return network.Ignore, nil
+	}
+
+	err := builder.Add(pos, sfa.Sig, true)
+	if err != nil {
+		return network.Disconnect, err
+	}
+
+	err = ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return addPendingSig(tx, sfa.Round, pendingSig{
+			signer:       sfa.Signer,
+			sig:          sfa.Sig,
+			fromThisNode: sender == nil,
+		})
+	})
+	if err != nil {
+		return network.Ignore, err
+	}
+
+	return network.Broadcast, nil
+}
+
+func (ccw *Worker) builder() {
```

Comment:
How do we handle catchup? Sounds like signatures are not needed during catchup on a blockchain that already has compact signatures

Reply:
There's two possible directions in terms of how catchup and compact certs could interact.

One direction is: compact certs could be used to significantly speed up catchup.  This is not done in this PR, but hopefully should be not too hard to implement.  The block validation logic can be borrowed from the compactcert e2e test, and used during catchup to quickly validate a chain of blocks at multiples of CompactCertRounds.  Then use the prev pointer (`BlockHdr.Branch`) to validate previous block hashes.

The other direction is: during catchup, the compact cert code will try to sign blocks for compact certs, and build compact certs.  If compact certs have already been formed for those blocks (most likely), this will be harmless.  But the code tries to be careful and make sure that, even if a node was not running at some point, restarting it will cause it to produce all the same compact cert signatures, in case the network was for some reason waiting for our signature.